source/mir/random/tinflex/package.d

```diff
@@ -0,0 +1,207 @@
+    "$(LINK2 http://epub.wu-wien.ac.at/3158/1/techreport-110.pdf,
+    Transformed density rejection with inflection points.)"
+    Statistics and Computing 23.2 (2013): 251-260.
+
+Macros:
+    F_TILDE=g(x)
+*/
+module mir.random.tinflex;
+
+import mir.random.tinflex.internal.types : IntervalPoint;
+
+import std.traits : ReturnType;
+import std.random : isUniformRNG;
+
+
+/**
+The Transformed Density Rejection with Inflection Points (Tinflex) algorithm
+can sample from arbitrary distributions given its density function f, its
+first two derivatives and a partitioning into intervals with at most one inflection
+point.
+
+Params:
+    f0 = probability density function of the distribution
+    f1 = first derivative of f0
+    f2 = second derivative of f0
+    c = T_c family
+    points = non-overlapping partitioning with at most one inflection point per interval
+    rho = efficiency of the Tinflex algorithm
+
+Returns:
+    Tinflex Generator.
+*/
+Tinflex!(F0, S) tinflex(F0, F1, F2, S)
+               (in F0 f0, in F1 f1, in F2 f2,
+                S c, S[] points, S rho = 1.1)
+{
+    import mir.random.tinflex.internal.calc : calcPoints;
+    // pre-calculate all the points
+    auto ips = calcPoints(f0, f1, f2, c, points, 1.1);
+    return Tinflex!(F0, S)(f0, ips, c);
+}
+
+/**
+Data body of the Tinflex algorithm.
+Can be used to sample from the distribution.
+*/
+struct Tinflex(F0, S)
+{
+    // density function
+    private const F0 _f0;
+
+    // generated partition points
+    private const IntervalPoint!S[] _ips;
+
+    // global T_c family
+    private const S c;
+
+    private this(const F0 f0, const IntervalPoint!S[] ips, const S c)
+    {
+        _f0 = f0;
+        _ips = ips;
+        this.c = c;
+    }
+
+    /// density function of the distribution
+    F0 pdf() const @property
+    {
+        return _f0;
+    }
+
+    /// density function of the distribution
+    S pdf(S x) const @property
+    {
+        return _f0(x);
+    }
+
+    /// Generated partition points
+    const(IntervalPoint!S[]) ips() @property const
+    {
+        return _ips;
+    }
+
+    /**
+    Sample n times from the distribution.
+    Params:
+        n = number of times to samples
+        rng = random number generator to use
+    Returns:
+        Array of length n with the samples
+    */
+    S opCall() const
+    {
+        import std.random : rndGen;
+        return tfSample(_f0, _ips, c, rndGen);
+    }
+
+    /// ditto
+    S opCall(RNG)(ref RNG rng) const
+        if (isUniformRNG!RNG)
+    {
+        return tfSample(_f0, _ips, c, rng);
+    }
+}
+
+/**
+Sample from the distribution.
+Params:
+    ips = calculated inflection points
+    rng = random number generator to use
+*/
+protected S tfSample(F0, S, RNG)
+          (in F0 f0, in IntervalPoint!S[] ips, in S c, ref RNG rng)
+    if (is(ReturnType!F0 == S) &&
+        isUniformRNG!RNG)
+{
+    import std.algorithm: filter, joiner, map, sum;
```

Comment:
By proper design this function should not have any code that can use something from `std.algorithm`. Is needs only binary search from `std.range.SortedRange`.